pkg/privilege/privileges/cache.go

```diff
@@ -1633,12 +1815,40 @@ func (p *MySQLPrivilege) getAllRoles(user, host string) []*auth.RoleIdentity {
 
 // Handle wraps MySQLPrivilege providing thread safe access.
 type Handle struct {
+	sctx sessionctx.Context
 	priv atomic.Pointer[MySQLPrivilege]
+	// Only load the active user's data to save memory
+	// username => struct{}
+	activeUsers sync.Map
 }
 
 // NewHandle returns a Handle.
-func NewHandle() *Handle {
-	return &Handle{}
+func NewHandle(sctx sessionctx.Context) *Handle {
+	var priv MySQLPrivilege
+	ret := &Handle{}
+	ret.sctx = sctx
+	ret.priv.Store(&priv)
+	return ret
+}
+
+// ensureActiveUser ensure that the specific user data is loaded in-memory.
+func (h *Handle) ensureActiveUser(user string) error {
+	_, exist := h.activeUsers.Load(user)
+	if exist {
+		return nil
+	}
+
+	var data immutable
+	err := data.loadSomeUsers(h.sctx, user)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	old := h.Get()
+	h.priv.Store(old.merge(&data))
```

Comment:
seems `ensureActiveUser` is concurrently used? For 2 concurrent call with different `user` and this timeline

1. user 1 is not active, then `loadSomeUsers` for it, then `h.Get()`
2. user 2 is not active, then `loadSomeUsers` for it, then `h.Get()`
3. `h.priv.Store` saved the merge result with user 1, and mark user 1 active
4. `h.priv.Store` saved the merge result with user 2 but data from user 1 is lost, and mark user 2 active

Now both users are active, but the the meaning of "ensure" is not achieved

Reply:
Then we need either `lock` or `cas + version` semantic.

Reply:
Can we design a per-user `priv` instead of current all-users`priv` object?

Reply:
Each session handle their own login user's privileges?

Reply:
> Then we need either lock or cas + version semantic.

Now changed to CAS @lance6716